packages/platform-android/native_modules.gradle

```diff
@@ -126,11 +126,37 @@ class ReactNativeModules {
     String packageClassInstances = ""
 
     if (packages.size() > 0) {
-      packageImports = "import ${packageName}.BuildConfig;\nimport ${packageName}.R;\n\n"
-      packageImports = packageImports + packages.collect {
-        "// ${it.name}\n${it.packageImportPath}"
+      def interpolateDynamicValues = {
+        it
+                // Before adding the {{ packageName }} replacement mechanism,
+                // BuildConfig and R classes were imported automatically
+                // into the scope of the file. In order to be backwards-compatible
+                // let's prefix all non-FQDN references to those classes with
+                // the package name.
+                //
+                // We want to match:
+                //  - new Package(R.string…),
```

Comment:
So I guess the essence of this PR is that: "if automatic prefixing doesn't work - produces an error, please use overrides and set {{packageName}} yourself"? 

If we consider this deprecated, we should probably print a warning that this will stop working in the future.

If we consider this backward-compatible support something that will stay for longer, then I would say we should document the exact use-case in the documentation - I've seen a comment from you to @Salakar, but being not that proficient with Android, I have trouble wrapping my head around it fully. In what cases this is going to break?

Reply:
I'm not sure if I have understood your comment correctly, so let me rephrase the essence:

> if the library doesn't use `{{ packageName }}` template, but it looks like it expects `BuildConfig` and `R` to be imported globally, prefix their usages with the template

This way we:
- keep the solution backwards-compatible,
- if for some reason a developer would expect to be able to access some other `packageName`-dependant class, they will be able to use `{{ packageName }}` instead of making a pull request to add another `import {{ packageName }}.AnotherClass` to the top of the file.

I will explain the exact use case by the exact use case that resulted in this PR 😃 

In Expo repository our Android project currently looks like this:

- `android/app` — where `host.exp.exponent.MainApplication` lives
- `android/expoview` — where the `host.exp.expoview` library lives, a monolith native Expo library containing vendored native modules (eg. `react-native-screens`)
- `packages/…` — unimodules projects

`app` depends on `expoview`, `expoview` depends on unimodules. It doesn't need to depend on React Native modules, since all of them are copied into `expoview`.

What I wanted to do was to replace copied RN modules with autolinking (from `/node_modules`).

In order to do this I added the autolinking script to the `expoview` project. The generated `PackageList` contained `import host.exp.exponent.R;`. Notice the `host.exp.exponent` — it has been inferred from the package of the `MainApplication`, so a different Android project. 

On one hand it makes sense, given that `MainApplication` is in `host.exp.exponent` package; on the other hand, this `PackageList` was being used in `host.exp.expoview` package, so the appropriate `R` class would be `host.exp.expoview.R` package. This mismatch resulted in a compilation error, since the mispackaged class could not have been found.

This particular PR won't solve this use case fully (if we wanted to depend on CodePush in `expoview`, it would still import `host.exp.exponent.BuildConfig`, so the proper fix would also include a different strategy for fetching `packageName`), but it's good enough for us and other multi-library Android projects (+ I suspect the `packageName` strategy fix would a major refactor).

To sum up, this will allow using React Native autolinking in Android workspaces where one or more applications-projects depend on an internal Android library-project which autolinks React Native modules. 🙂

Reply:
Before I dive deep, let's clarify one thing (I need to ask fewer questions at once, haha). Happy to hop on a call as well in case it's easier.

> if the library doesn't use {{ packageName }} template, but it looks like it expects BuildConfig and R to be imported globally, prefix their usages with the template

What is the benefit of doing a custom property on a CLI config like this: `new AwesomePackage({{ packageName }}.R.string.key)` if I can just leave do `new AwesomePackage(R.string.key)` and the backward-compatible mechanism will take care of prefixes?

Reply:
None! 😅  Maybe the PR’s essence should be rephrased to something along the lines of

> only use `{{ packageName }}.{R,BuildConfig}` if they are expected

which can also be read as

> do not import `{{ packageName }}.{R,BuildConfig}` if they aren’t used

and it could have been done like:

1. see if any of the packages uses R or BuildConfig class
2. if they do, add respective imports to the imports section of the file

I thought adding a placeholder would make it easier to maintain (if for whatever reason any package would need any other packageName-dependant class, it wouldn’t require any CLI change).